test/Datadog.Trace.ClrProfiler.IntegrationTests/Helpers/SpanExpectation.cs

```diff
@@ -0,0 +1,211 @@
+            {
+                var mismatchMessage = assertion(span);
+                if (!string.IsNullOrWhiteSpace(mismatchMessage))
+                {
+                    messages.Add(mismatchMessage);
+                }
+            }
+
+            if (messages.Any())
+            {
+                message = string.Join(",", messages);
+                return false;
+            }
+
+            return true;
+        }
+
+        public void TagShouldExist(string tagKey, Func<MockTracerAgent.Span, bool> when)
+        {
+            Assertions.Add(span =>
+            {
+                if (when(span) && !span.Tags.ContainsKey(tagKey))
+                {
+                    return $"Tag {tagKey} is missing from span.";
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterDelegateExpectation(Func<MockTracerAgent.Span, IEnumerable<string>> expectation)
+        {
+            Assertions.Add(span =>
+            {
+                var failures = expectation(span);
+
+                if (failures.Any())
+                {
+                    return string.Join(",", failures);
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterCustomExpectation(string keyForMessage, Func<MockTracerAgent.Span, string> actual, Func<MockTracerAgent.Span, string> expected)
+        {
+            Assertions.Add(span =>
+            {
+                var actualValue = actual(span);
+                var expectedValue = expected(span);
+
+                if (expectedValue != null && actualValue != expectedValue)
+                {
+                    return FailureMessage(name: keyForMessage, actual: actualValue, expected: expectedValue);
+                }
+
+                return null;
+            });
+        }
+
+        public void RegisterTagExpectation(string key, string expected, Func<MockTracerAgent.Span, bool> when, Func<string, string> customMessage = null)
+        {
+            Assertions.Add(span =>
+            {
+                if (!when(span))
+                {
+                    return null;
+                }
+
+                var actualValue = GetTag(span, key);
+
+                if (expected != null && actualValue != expected)
+                {
+                    if (customMessage != null)
+                    {
+                        return customMessage(actualValue);
+                    }
+
+                    return FailureMessage(name: key, actual: actualValue, expected: expected);
+                }
+
+                return null;
+            });
+        }
+
+        protected string FailureMessage(string name, string actual, string expected)
+        {
+            return $"({name} mismatch: actual: {actual ?? "NULL"}, expected: {expected ?? "NULL"})";
+        }
+
+        protected string GetTag(MockTracerAgent.Span span, string tag)
+        {
+            if (span.Tags.ContainsKey(tag))
```

Comment:
nit: This doesn't need any changes, but I always support using `TryGetValue` when possible, which could condense these lines to:

```
span.Tags.TryGetValue(tag, out string value);
return value;
```

Reply:
We also have a helper `GetValueOrDefault()` extension method if you don't like `out` parameters.